torch/optim/sgd.py

```diff
@@ -349,8 +355,10 @@ def _single_tensor_sgd(
                 grad = grad.add(buf, alpha=momentum)
             else:
                 grad = buf
-
-        param.add_(grad, alpha=-lr)
+        if differentiable and isinstance(lr, Tensor) and lr.requires_grad:
+            param.addcmul_(grad, -lr.to(param.device))
```

Comment:
This `.to` is expensive...I'm trying to brainstorm the best way to do this and will get back to you tomorrow.

My thoughts so far: if lr is on CPU, we should be able to take advantage of its CPUness and just .item() before passing it to CUDA, to avoid this expensive `.to` op. If that's not easy to do quickly (I haven't figured out why addcmul doesn't automatically allow it yet), we should maintain a lr_dict similar to in the fused adam optimizer. I'll get back to you!

Reply:
Alright! So here's the ideal path forward for performance:

1. We _should_ figure out how to take advantage of the CPUness of the LR and just .item() before passing it to CUDA for addcmul. Note that the following works for some torch ops already (add and mul):
```
import torch
grad = torch.rand(2, 4, device="cuda")
lr = torch.tensor(1e-3, requires_grad=True)

torch.add(grad, lr)   # works!
torch.mul(grad, lr)   # also works!

torch.addcmul(grad, grad, lr)  # does not work :/, device mismatch
```
Step one would be to write a PR (with a test) that enables that line to work just like add and mul work. The CUDA kernel for addcmul is in PointwiseOpsKernel.cu.

2. Then come back to this PR and use `param.addcmul_(grad, lr, -1)`